include/ensmallen_bits/callbacks/grad_clip_by_value.hpp

```diff
@@ -0,0 +1,61 @@
+/**
+ * @file grad_clip_by_value.hpp
+ * @author Marcus Edel
+ *
+ * Clips the gradient to a specified min and max.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_CALLBACKS_GRAD_CLIP_BY_VALUE_HPP
+#define ENSMALLEN_CALLBACKS_GRAD_CLIP_BY_VALUE_HPP
+
+namespace ens {
+
+/**
+ * Clip the gradient to a specified min and max.
+ */
+class GradClipByValue
+{
+ public:
+  /**
+   * Set up the gradient clip by value callback class with the min and max
+   * value.
+   *
+   * @param min The minimum value to clip to.
+   * @param max The maximum value to clip to.
+   */
+  GradClipByValue(const double min, const double max) : lower(min), upper(max)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Callback function called at any call to Gradient().
+   *
+   * @param optimizer The optimizer used to update the function.
+   * @param function Function to optimize.
+   * @param coordinates Starting point.
+   * @param gradient Matrix that holds the gradient.
+   */
+  template<typename OptimizerType, typename FunctionType, typename MatType>
+  void Gradient(OptimizerType& /* optimizer */,
+                FunctionType& /* function */,
+                const MatType& /* coordinates */,
+                MatType& gradient)
+  {
+    for (size_t i = 0; i < gradient.n_elem; ++i)
+      gradient(i) = std::max(lower, std::min(gradient(i), upper));
```

Comment:
Maybe you can use `clamp()` here instead?  http://arma.sourceforge.net/docs.html#clamp_member

Reply:
Ahh, yeah, I was looking if there is an existing armadillo function, but searched for the wrong keyword.

Reply:
Quick note on `.clamp()` vs `clamp()`.  The member function [`.clamp()`](http://arma.sourceforge.net/docs.html#clamp_member) is in Armadillo 10.5 onwards. The stand-alone function [`clamp()`](http://arma.sourceforge.net/docs.html#clamp) is in Armadillo 4.400 onwards.  It was expanded to handle cubes in 7.900, and sparse matrices in 9.600.

Currently the minimum Armadillo version required by ensmallen in 8.400.  I recommend bumping this to 9.800, which is in Ubuntu 20.04 LTS.  This will allow `clamp()` to be used by sparse matrices within ensmallen.

20.04 LTS has been out for more than a year, which is plenty of time for people to upgrade from 18.04.  In any case it's quite straightforward to manually update Armadillo to the latest version.

Reply:
@conradsnicta had left a comment that pointed out that `.clamp()` wasn't available until recently, so we should probably use the standalone `clamp()` version.  But there is also another observation, which is that `clamp()` did not work on sparse matrices until 9.600.  I think it's not worth updating the minimum version of Armadillo for this (since I'd like to keep it in sync with mlpack's minimum required version of Armadillo), but, it might be worth special-casing this:

```
  #if ARMA_VERSION_MAJOR < 9 || ARMA_VERSION_MAJOR == 9 && ARMA_VERSION_MINOR < 600
    if (arma::is_arma_sparse_type<MatType>::value)
      // use manually implemented loop
    else
      // use clamp()
  #else
    // use clamp()
  #endif
```
  
and then when the time to upgrade the minimum version comes, we can remove this part.  (Alternately, we could add a caveat to the documentation pointing out that if you are on Armadillo < 9.600 and using sparse gradients, that this callback will not work.  The set of people realistically interested in doing that is probably vanishingly small.)

Reply:
I'm certainly not keen on having `#if`s throughout the codebase, especially with backported or manually implemented workarounds.  From first hand experience this type of approach is polluting and considerably increases the maintenance burden.  Little bit here and there quickly adds up, and requires tracking (either explicit or implicit).

I don't think we should aim to accommodate everybody's setup -- it's too much work in the end for something that is volunteer driven.  There are only so many hours in the day, and we all have other things on our plates.  It would be far less hassle to simply encourage people to upgrade their version of Armadillo if they want to use sparse gradients.

I'd wager that anybody sufficiently sophisticated to use sparse gradients would know how to update Armadillo on their system.

So overall, I'm strongly in favour of simply adding a caveat to the documentation, with the minimum suggested version of Armadillo at 9.800 (which is in Ubuntu 20.04 LTS).

Reply:
Yeah, agreed---I think the note in the documentation is probably the way to go here then.

Reply:
Sounds reasonable to me, I can open another PR with the necessary changes.

Reply:
Resolved by #318